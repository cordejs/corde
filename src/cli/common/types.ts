import { Command } from "commander";

export type ParamsFrom = "args" | "options" | "both";

export interface ICommandOptions {
  flags: string;
  description?: string;
  defaultValue?: string;
}

export interface ICliCommandConstructor {
  name?: string;
  program: Command;
  paramsFrom?: ParamsFrom;
}

export interface ICliCommand {
  readonly paramsFrom?: ParamsFrom;
  readonly command: Command;
  handler(...args: any[]): Promise<void> | void;
}

export interface ICliCommandConstructable {
  new (command: Command): ICliCommand;
}
