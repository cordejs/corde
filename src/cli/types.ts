export interface ICommand {
  action(...args: any[]): Promise<void> | void;
}
