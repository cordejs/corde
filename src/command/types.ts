import { ICordeBot } from "../types";

export interface ICreateMatcherParam {
  matcher?: string;
  isNot?: boolean;
  commandName?: string | boolean | number;
  channelId?: string;
  guildId?: string;
  isDebug?: boolean;
  isCascade?: boolean;
  cordeBot?: ICordeBot;
  trace?: string;
  utility?: boolean;
  mustSendCommand?: boolean;
}
