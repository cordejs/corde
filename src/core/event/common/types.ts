import { Client, ClientEvents } from "discord.js";

export type EventNames = keyof ClientEvents;

export type Spread<T extends EventNames> = ClientEvents[T][1] extends undefined
  ? ClientEvents[T] extends (infer U)[]
    ? U
    : never
  : ClientEvents[T];

export interface IDiscordEvent<TEvent extends EventNames, TFilter> {
  canListen(): boolean;
  getIntents(): string[];
  on(fn: (...args: ClientEvents[TEvent]) => void): this;
  once(filter?: TFilter): Promise<Spread<TEvent>>;
}

export interface IDiscordEventConstructable<TEvent extends EventNames, TFilter> {
  new (client: Client): IDiscordEvent<TEvent, TFilter>;
}

export type IntentType =
  | "GUILDS"
  | "GUILD_MEMBERS"
  | "GUILD_BANS"
  | "GUILD_EMOJIS_AND_STICKERS"
  | "GUILD_INTEGRATIONS"
  | "GUILD_WEBHOOKS"
  | "GUILD_INVITES"
  | "GUILD_VOICE_STATES"
  | "GUILD_PRESENCES"
  | "GUILD_MESSAGES"
  | "GUILD_MESSAGE_REACTIONS"
  | "GUILD_MESSAGE_TYPING"
  | "DIRECT_MESSAGES"
  | "DIRECT_MESSAGE_REACTIONS"
  | "DIRECT_MESSAGE_TYPING"
  | "GUILD_SCHEDULED_EVENTS";

export type EventClass = {
  [k in IntentType]: {
    bitfield: number;
    events: EventNames[];
  };
};
