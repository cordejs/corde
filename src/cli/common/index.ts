export * from "./loadConfigs";
export * from "./commandFactory";
