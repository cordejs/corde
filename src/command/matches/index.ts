export * from "./message";
export * from "./role";
