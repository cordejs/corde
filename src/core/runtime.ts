import { ICordeBot } from "../types";
import { IInternalEvents } from "./internalEvents";
import { CordeBot } from "./CordeBot";
import { Client, Intents } from "discord.js";
import { Config } from "./Config";
import { TestCollector } from "./TestCollector";
import { ConfigError } from "../errors";
import path from "path";
import { replaceAll } from "../utils/replaceAll";
import { ROOT_DIR } from "../const";
import { EventEmitter } from "stream";

/**
 * @internal
 */
export class Runtime {
  public configFilePath!: string;
  public files!: string[];

  public readonly testCollector: TestCollector;
  private _internalEvents: IInternalEvents;
  private _mocks: Array<corde.IMockInstance<any, any, any>>;

  private readonly _configs: Config;
  private _bot!: ICordeBot;

  get bot() {
    return this._bot;
  }

  get isUnityTest() {
    return process.env.ENV === "UNITY_TEST";
  }

  get isE2eTest() {
    return process.env.ENV === "E2E_TEST";
  }

  get isTestEnv() {
    return this.isE2eTest || this.isUnityTest;
  }

  get exitOnFileReadingError() {
    return this._configs.exitOnFileReadingError;
  }

  get extensions() {
    return this._configs.extensions;
  }

  get events() {
    return this.bot.events;
  }

  get configs() {
    return this._configs;
  }

  get cordeBotToken() {
    return this._configs.cordeBotToken;
  }

  get botTestId() {
    return this._configs.botTestId;
  }

  get project() {
    return this._configs.project;
  }

  get channelId() {
    return this._configs.channelId;
  }

  get guildId() {
    return this._configs.guildId;
  }

  get timeout() {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return this._configs.timeout!;
  }

  get botPrefix() {
    return this._configs.botPrefix;
  }

  get testMatches() {
    return this._configs.testMatches;
  }

  get modulePathIgnorePatterns() {
    return this._configs.modulePathIgnorePatterns;
  }

  get internalEvents() {
    return this._internalEvents;
  }

  constructor() {
    this._internalEvents = new EventEmitter();
    this._configs = new Config();
    this._mocks = [];
    this.testCollector = new TestCollector();
  }

  setConfigs(_configs: Partial<corde.IConfigOptions>, forceUpdate?: boolean) {
    if (!_configs) {
      throw new ConfigError("Invalid _configs");
    }

    this._configs.setConfigs(_configs, forceUpdate);
  }

  replaceWithRootDir(text: string) {
    if (this._configs.rootDir) {
      return replaceAll(text, ROOT_DIR, this._configs.rootDir);
    }
    return text;
  }

  resolvePathWithRootDir(partialPath: string) {
    if (this._configs.rootDir) {
      return path.resolve(process.cwd(), this._configs.rootDir, partialPath);
    }
    return path.resolve(process.cwd(), partialPath);
  }

  /**
   * Shortcut for *bot.isLoggedIn*
   */
  isBotLoggedIn() {
    return this.bot && this.bot.isLoggedIn();
  }

  /**
   * Shortcut for *bot.logout*
   */
  logoffBot() {
    if (this._bot) {
      this._bot.logout();
    }
  }

  loginBot(token: string) {
    return this.bot.login(token);
  }

  initBot() {
    this._bot = new CordeBot(
      this._configs.botPrefix,
      this._configs.guildId,
      this._configs.channelId,
      this._configs.botTestId,
      this.initClient(),
    );
  }

  addMock(mock: corde.IMockInstance<any, any, any>) {
    this._mocks.push(mock);
  }

  resetAllMocks() {
    this._mocks.forEach((mock) => mock.restore());
  }

  private initClient() {
    return new Client({
      intents: [Intents.FLAGS.GUILDS],
    });
  }
}

const runtime = new Runtime();
export default runtime;
