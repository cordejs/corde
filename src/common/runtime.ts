import { IConfigOptions, ICordeBot, TestFunctionType } from "../types";
import { Config } from "./config";
import { Client } from "discord.js";
import { CordeBot } from "../core/cordeBot";
import { ConfigError } from "../errors";

const Environment = {
  isUnityTest: process.env.ENV === "UNITY_TEST",
  isE2eTest: process.env.ENV === "E2E_TEST",
};

/**
 * @internal
 */
class Runtime implements IConfigOptions {
  get bot() {
    if (!this._bot) {
      this._bot = this.initBot();
    }

    return this._bot;
  }

  get isTestEnv() {
    return this.environment.isE2eTest || this.environment.isUnityTest;
  }

  get environment() {
    return Environment;
  }

  get exitOnFileReadingError() {
    return this._configs.exitOnFileReadingError;
  }

  get extensions() {
    return this._configs.extentions;
  }

  get events() {
    return this.bot.events;
  }

  get configs() {
    return this._configs;
  }

  get cordeBotToken() {
    return this._configs.cordeBotToken;
  }

  get botTestId() {
    return this._configs.botTestId;
  }

  get project() {
    return this._configs.project;
  }

  get botToken() {
    return this._configs.botToken;
  }

  get channelId() {
    return this._configs.channelId;
  }

  get guildId() {
    return this._configs.guildId;
  }

  get timeout() {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return this._configs.timeout!;
  }

  get botPrefix() {
    return this._configs.botPrefix;
  }

  get testMatches() {
    return this._configs.testMatches;
  }

  get modulePathIgnorePatterns() {
    return this._configs.modulePathIgnorePatterns;
  }

  private constructor() {
    this._configs = new Config();
  }

  private static _instance: Runtime;
  configFilePath!: string;
  files!: string[];

  private readonly _configs: Config;
  private _bot!: ICordeBot;

  static getInstance() {
    if (!Runtime._instance) {
      Runtime._instance = new Runtime();
    }
    return Runtime._instance;
  }

  setConfigs(_configs: Partial<IConfigOptions>, forceUpdate?: boolean) {
    if (!_configs) {
      throw new ConfigError("Invalid _configs");
    }

    this._configs.setConfigs(_configs, forceUpdate);
  }

  /**
   * Shortcut for *bot.isLoggedIn*
   */
  isBotLoggedIn() {
    return this.bot && this.bot.isLoggedIn();
  }

  /**
   * Shortcut for *bot.logout*
   */
  logoffBot() {
    if (this._bot) {
      this._bot.logout();
    }
  }

  async loginBot(token: string) {
    return await this.bot.login(token);
  }

  injectBot(fn: TestFunctionType) {
    return fn(this.bot);
  }

  initBot() {
    return new CordeBot(
      this._configs.botPrefix,
      this._configs.guildId,
      this._configs.channelId,
      this._configs.botTestId,
      new Client(),
    );
  }
}

/**
 * Singleton of Runtime.
 */
const runtime = Runtime.getInstance();
export { runtime };
