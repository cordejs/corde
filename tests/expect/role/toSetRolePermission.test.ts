import { Client } from "discord.js";
import { ToSetRolePermission } from "../../../src/expect/matches";
import MockDiscord from "../../mocks/mockDiscord";
import {
  createReport,
  initCordeClientWithChannel,
  removeANSIColorStyle,
  testUtils,
} from "../../testHelper";
import { ICordeBot, ITestReport } from "../../../src/types";
import {
  buildReportMessage,
  calcPermissionsValue,
  diff,
  Permission,
  permissionsArray,
} from "../../../src/utils";
import { MockEvents } from "../../mocks/mockEvents";
import { runtime } from "../../../src/common/runtime";

let mockDiscord = new MockDiscord();

function initClient() {
  const corde = initCordeClientWithChannel(mockDiscord, new Client());
  corde.findRole = jest.fn().mockReturnValue(mockDiscord.role);
  corde.fetchRole = jest.fn().mockReturnValue(mockDiscord.role);
  corde.sendTextMessage = jest.fn().mockImplementation(() => {});
  return corde;
}

function initTestClass(cordeBot: ICordeBot, isNot: boolean) {
  return testUtils.initTestClass(ToSetRolePermission, {
    command: "toDelete",
    cordeBot: cordeBot,
    isNot: isNot,
  });
}

describe("testing toSetRolePermission operation", () => {
  afterEach(() => {
    mockDiscord = new MockDiscord();
  });

  it("should fail due to undefined roleIdentifier", async () => {
    const corde = initClient();
    const toSetRolePermission = initTestClass(corde, false);
    const report = await toSetRolePermission.action(undefined, ["ADD_REACTIONS"]);

    const message = buildReportMessage(
      "expected: data to identifier the role (id or name)\n",
      `received: null`,
    );

    const expectReport = createReport(toSetRolePermission, false, message);

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to invalid permission parameter (object)", async () => {
    const corde = initClient();
    const toSetRolePermission = initTestClass(corde, false);
    // @ts-ignore
    const report = await toSetRolePermission.action({ id: "123" }, {});

    const message = buildReportMessage(
      `expected: permissions to be null, undefined or an array\n`,
      `received: object`,
    );

    const expectReport = createReport(toSetRolePermission, false, message);

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to invalid permission value", async () => {
    const corde = initClient();
    const toSetRolePermission = initTestClass(corde, false);
    // @ts-ignore
    const report = await toSetRolePermission.action({ id: "123" }, ["BANANA"]);

    const message = buildReportMessage(diff(permissionsArray, ["BANANA"]));

    const expectReport = createReport(toSetRolePermission, false, message);

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to invalid permission parameter (undefined)", async () => {
    const corde = initClient();
    const toSetRolePermission = initTestClass(corde, false);
    // @ts-ignore
    const report = await toSetRolePermission.action({ id: "123" }, {});

    const message = buildReportMessage(
      `expected: permissions to be null, undefined or an array\n`,
      `received: object`,
    );

    const expectReport = createReport(toSetRolePermission, false, message);

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to not found role", async () => {
    const corde = initClient();
    corde.findRole = jest.fn().mockReturnValue(null);
    const toSetRolePermission = initTestClass(corde, false);
    const report = await toSetRolePermission.action({ id: "123" }, ["ATTACH_FILES"]);

    const message = buildReportMessage(`expected: role with id 123\n`, `received: null`);

    const expectReport = createReport(toSetRolePermission, false, message);

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to role permissions was not updated", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);

    const toSetRolePermission = initTestClass(corde, false);
    const report = await toSetRolePermission.action({ id: "123" }, ["ATTACH_FILES"]);

    const message = buildReportMessage(
      `expected: role permissions change to: ATTACH_FILES\n`,
      `received: permissions were not changed`,
    );

    const expectReport = createReport(toSetRolePermission, false, message);

    expect(report).toEqual(expectReport);
    const snapReport = {
      pass: report.pass,
      message: removeANSIColorStyle(report.message),
    };
    expect(snapReport).toMatchSnapshot();
  });

  it("should return true due to isNot true and no role change", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);

    const toSetRolePermission = initTestClass(corde, true);
    const report = await toSetRolePermission.action({ id: "123" }, ["ATTACH_FILES"]);

    const expectReport = createReport(toSetRolePermission, true);

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return true due to role changed permissions (isNot false)", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);
    const mockRole = mockDiscord.createMockRole(
      "test role",
      calcPermissionsValue(...mockDiscord.role.permissions.toArray().map((p) => Permission[p])),
    );
    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const toSetRolePermission = initTestClass(corde, false);
    const report = await toSetRolePermission.action({ id: "123" }, mockRole.permissions.toArray());

    const expectReport = createReport(toSetRolePermission, true);

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due to permissions should not change (isNot true)", async () => {
    const corde = initClient();
    const mockRole = mockDiscord.createMockRole(
      "test role",
      calcPermissionsValue(Permission.ADMINISTRATOR),
    );
    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);

    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const toSetRolePermission = initTestClass(corde, true);
    const report = await toSetRolePermission.action({ id: "123" }, ["ADMINISTRATOR"]);

    const message = buildReportMessage(
      `expected: role permissions not change to: ADMINISTRATOR\n`,
      `received: ADMINISTRATOR (and ${
        mockRole.permissions.toArray().filter((p) => p !== "ADMINISTRATOR").length
      } others)`,
    );

    const expectReport = createReport(toSetRolePermission, false, message);

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due to permissions should not change (isNot true)", async () => {
    const corde = initClient();
    const mockRole = mockDiscord.createMockRole(
      "test role",
      calcPermissionsValue(Permission.ADMINISTRATOR, Permission.BAN_MEMBERS),
    );
    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);

    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const toSetRolePermission = initTestClass(corde, true);
    const report = await toSetRolePermission.action({ id: "123" }, [
      "ADMINISTRATOR",
      "BAN_MEMBERS",
    ]);

    const message = buildReportMessage(
      `expected: role permissions not change to: ADMINISTRATOR and BAN_MEMBERS\n`,
      `received: ADMINISTRATOR (and ${
        mockRole.permissions.toArray().filter((p) => p !== "ADMINISTRATOR").length
      } others)`,
    );

    const expectReport = createReport(toSetRolePermission, false, message);

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due to permissions should not change (isNot true)", async () => {
    const corde = initClient();
    const mockRole = mockDiscord.createMockRole(
      "test role",
      calcPermissionsValue(Permission.ADMINISTRATOR, Permission.BAN_MEMBERS, Permission.CONNECT),
    );
    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);

    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const toSetRolePermission = initTestClass(corde, true);
    const report = await toSetRolePermission.action({ id: "123" }, [
      "ADMINISTRATOR",
      "BAN_MEMBERS",
      "CONNECT",
    ]);

    const message = buildReportMessage(
      `expected: role permissions not change to: ADMINISTRATOR (and 2 others)\n`,
      `received: ADMINISTRATOR (and ${
        mockRole.permissions.toArray().filter((p) => p !== "ADMINISTRATOR").length
      } others)`,
    );

    const expectReport: ITestReport = {
      pass: false,
      message,
      testName: toSetRolePermission.toString(),
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due expected name did not match to received", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);
    mockEvent.mockOnceRolePermissionsUpdate(mockDiscord.role);
    const toSetRolePermission = initTestClass(corde, false);
    const report = await toSetRolePermission.action({ id: "123" }, ["ATTACH_FILES"]);

    const message = buildReportMessage(
      `expected: role permissions change to: ATTACH_FILES\n`,
      `received: ADMINISTRATOR (and 30 others)`,
    );

    const expectReport: ITestReport = {
      pass: false,
      message,
      testName: toSetRolePermission.toString(),
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due expected name did not match to received", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);
    mockEvent.mockOnceRolePermissionsUpdate(mockDiscord.role);
    const toSetRolePermission = initTestClass(corde, false);
    const report = await toSetRolePermission.action({ id: "123" }, null);

    const message = buildReportMessage(
      `expected: role permissions change to: null\n`,
      `received: ADMINISTRATOR (and 30 others)`,
    );

    const expectReport: ITestReport = {
      pass: false,
      message,
      testName: toSetRolePermission.toString(),
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due to failure in message sending", async () => {
    const corde = initCordeClientWithChannel(mockDiscord, mockDiscord.client);

    corde.findRole = jest.fn().mockReturnValue(mockDiscord.role);
    corde.fetchRole = jest.fn().mockReturnValue(null);

    const erroMessage = "can not send message to channel x";
    corde.sendTextMessage = jest
      .fn()
      .mockImplementation(() => Promise.reject(new Error(erroMessage)));

    const toSetRolePermission = initTestClass(corde, false);
    const report = await toSetRolePermission.action({ id: "123" }, ["ATTACH_FILES"]);

    const expectReport = createReport(toSetRolePermission, false, buildReportMessage(erroMessage));

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });
});
