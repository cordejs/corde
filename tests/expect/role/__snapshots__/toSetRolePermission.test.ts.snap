// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing toSetRolePermission operation should fail due to invalid permission value 1`] = `
Object {
  "message": "    
    - Expected  - 62
    + Received  +  1
    
      Array [
    -   \\"1\\",
    -   \\"2\\",
    -   \\"4\\",
    -   \\"8\\",
    -   \\"16\\",
    -   \\"32\\",
    -   \\"64\\",
    -   \\"128\\",
    -   \\"256\\",
    -   \\"512\\",
    -   \\"1024\\",
    -   \\"2048\\",
    -   \\"4096\\",
    -   \\"8192\\",
    -   \\"16384\\",
    -   \\"32768\\",
    -   \\"65536\\",
    -   \\"131072\\",
    -   \\"262144\\",
    -   \\"524288\\",
    -   \\"1048576\\",
    -   \\"2097152\\",
    -   \\"4194304\\",
    -   \\"8388608\\",
    -   \\"16777216\\",
    -   \\"33554432\\",
    -   \\"67108864\\",
    -   \\"134217728\\",
    -   \\"268435456\\",
    -   \\"536870912\\",
    -   \\"1073741824\\",
    -   \\"CREATE_INSTANT_INVITE\\",
    -   \\"KICK_MEMBERS\\",
    -   \\"BAN_MEMBERS\\",
    -   \\"ADMINISTRATOR\\",
    -   \\"MANAGE_CHANNELS\\",
    -   \\"MANAGE_GUILD\\",
    -   \\"ADD_REACTIONS\\",
    -   \\"VIEW_AUDIT_LOG\\",
    -   \\"VIEW_CHANNEL\\",
    -   \\"SEND_MESSAGES\\",
    -   \\"SEND_TTS_MESSAGES\\",
    -   \\"MANAGE_MESSAGES\\",
    -   \\"EMBED_LINKS\\",
    -   \\"ATTACH_FILES\\",
    -   \\"READ_MESSAGE_HISTORY\\",
    -   \\"MENTION_EVERYONE\\",
    -   \\"USE_EXTERNAL_EMOJIS\\",
    -   \\"CONNECT\\",
    -   \\"SPEAK\\",
    -   \\"MUTE_MEMBERS\\",
    -   \\"DEAFEN_MEMBERS\\",
    -   \\"MOVE_MEMBERS\\",
    -   \\"USE_VAD\\",
    -   \\"PRIORITY_SPEAKER\\",
    -   \\"CHANGE_NICKNAME\\",
    -   \\"MANAGE_NICKNAMES\\",
    -   \\"MANAGE_ROLES\\",
    -   \\"MANAGE_WEBHOOKS\\",
    -   \\"MANAGE_EMOJIS\\",
    -   \\"STREAM\\",
    -   \\"VIEW_GUILD_INSIGHTS\\",
    +   \\"BANANA\\",
      ]
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should fail due to role permissions was not updated 1`] = `
Object {
  "message": "    
    expected: role permissions change to: ATTACH_FILES
    received: permissions were not changed
",
  "pass": false,
}
`;

exports[`testing toSetRolePermission operation should fail due to undefined roleIdentifier 1`] = `
Object {
  "message": "    
    expected: data to identifier the role (id or name)
    received: null
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return a failed test due to failure in message sending 1`] = `
Object {
  "message": "    
    can not send message to channel x
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return a not passed test due expected name did not match to received 1`] = `
Object {
  "message": "    
    expected: role permissions change to: ATTACH_FILES
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return a not passed test due expected name did not match to received 2`] = `
Object {
  "message": "    
    expected: role permissions change to: null
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return a not passed test due to permissions should not change (isNot true) 1`] = `
Object {
  "message": "    
    expected: role permissions not change to: ADMINISTRATOR
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return a not passed test due to permissions should not change (isNot true) 2`] = `
Object {
  "message": "    
    expected: role permissions not change to: ADMINISTRATOR and BAN_MEMBERS
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return a not passed test due to permissions should not change (isNot true) 3`] = `
Object {
  "message": "    
    expected: role permissions not change to: ADMINISTRATOR (and 2 others)
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return false due to invalid permission parameter (object) 1`] = `
Object {
  "message": "    
    expected: permissions to be null, undefined or an array
    received: object
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return false due to invalid permission parameter (undefined) 1`] = `
Object {
  "message": "    
    expected: permissions to be null, undefined or an array
    received: object
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return false due to not found role 1`] = `
Object {
  "message": "    
    expected: role with id 123
    received: null
",
  "pass": false,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return true due to isNot true and no role change 1`] = `
Object {
  "pass": true,
  "testName": "toSetRolePermission",
}
`;

exports[`testing toSetRolePermission operation should return true due to role changed permissions (isNot false) 1`] = `
Object {
  "pass": true,
  "testName": "toSetRolePermission",
}
`;
